ports/alif/machine_pin.c

```diff
@@ -30,8 +30,70 @@
 #include "extmod/virtpin.h"
 #include "shared/runtime/mpirq.h"
 
+#ifndef MACHINE_PIN_NUM_VECTORS
+#define MACHINE_PIN_NUM_PORT_IO (8)
+#define MACHINE_PIN_NUM_VECTORS (16 * MACHINE_PIN_NUM_PORT_IO)
+#endif
+
+typedef struct _machine_pin_irq_obj_t {
+    mp_irq_obj_t base;
+    uint32_t flags;
+    uint32_t trigger;
+    IRQn_Type irq_num;
+    bool reserved; // for use by other drivers
+} machine_pin_irq_obj_t;
+
+// Defines a single GPIO IRQ handler
+#define DEFINE_GPIO_IRQ_HANDLER(pname, port, pin) \
+    void pname##_IRQ##pin##Handler(void) { \
+        size_t irq_idx = port * MACHINE_PIN_NUM_PORT_IO + pin; \
+        machine_pin_irq_obj_t *irq = MP_STATE_PORT(machine_pin_irq_obj[irq_idx]); \
+        machine_pin_obj_t *self = MP_OBJ_TO_PTR(irq->base.parent); \
+        gpio_interrupt_eoi(self->gpio, pin); \
+        mp_irq_handler(&irq->base); \
```

Comment:
Need to set `irg->flags` to the flags that triggered the IRQ, before calling `mp_irq_handler()`.  Python code should be able to retrieve the flags (rising or falling or both) using `irq.flags()`.